### ISSUE TITLE
Decoding to pointers with nil value

### ISSUE THREAD

**Issue by cv21** [mitchellh/mapstructure#97]
Mapstructure is very useful and great library, thank you!

But many gophers use pointer type for know whether value is absent, or whether not.
The problem that mapstructure could not properly decode map values into struct field that is pointer.

**Example:**
We have target struct

```go
type Result struct {
	T *time.Time
}
```

and have a decode hook, which try to parse *string* to **time.Time*

```go
func (f reflect.Type, t reflect.Type, data interface{}) (interface{}, error) {
	if t == reflect.TypeOf(&time.Time{}) && f.Kind() == reflect.String && data == "" {
		return nil, nil
	}
        
        ...

	return data, nil
}
```

After decoding
***I would like to expect***
```
type Result struct {
	T *time.Time // nil
}
```
***But I got***
```
type Result struct {
	T *time.Time // 0001-01-01 00:00:00 +0000 UTC
}
```

The reason is that *decodePtr* function creates zero value for *valElemType*
```go
if realVal.IsNil() || d.config.ZeroFields {
	realVal = reflect.New(valElemType)
}
```

**Question**
Maybe we wont create zero value if data is nil and ZeroFields is false?
Or we can create specific tag for fields that could have nil value after decoding.

**Comment by YannickB** [comment]
+1

**Comment by snowzach** [comment]
+1 this is causing us some headache as well. Merging this pull request would be great.

**Comment by mitchellh** [comment]
Merged #135 and #137. 135 contains a test case for this and 137 contains changes that resolve this.
